File: frontend/src/types/graph.ts
import { AnomalyItem } from './anomaly';

export type NodeType = 
  | 'ANOMALY'
  | 'SERVICE'
  | 'REGION'
  | 'RESOURCE'
  | 'USAGE'
  | 'TEAM'
  | 'PROJECT'
  | 'DEPLOYMENT'
  | 'EVENT';

export interface InvestigationNode {
  id: string;
  type: NodeType;
  label: string;
  metadata: Record<string, any>;
}

export interface InvestigationEdge {
  id: string;
  source: string;
  target: string;
  relationship: string;
  strength: number;
  evidence: string[];
}

export interface InvestigationGraphSummary {
  primary_service?: string;
  primary_resource?: string;
  primary_region?: string;
  primary_event?: string;
  strongest_signal: string;
  confidence: number;
  evidence_count: number;
  node_count: number;
  edge_count: number;
}

export interface InvestigationGraphResponse {
  anomaly: AnomalyItem;
  nodes: InvestigationNode[];
  edges: InvestigationEdge[];
  summary: InvestigationGraphSummary;
}

File: backend/app/schemas/graph.py
from enum import Enum
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field
from backend.app.schemas.anomaly import AnomalyItemSchema


class NodeType(str, Enum):
    ANOMALY = "ANOMALY"
    SERVICE = "SERVICE"
    REGION = "REGION"
    RESOURCE = "RESOURCE"
    USAGE = "USAGE"
    TEAM = "TEAM"
    PROJECT = "PROJECT"
    DEPLOYMENT = "DEPLOYMENT"
    EVENT = "EVENT"


class InvestigationNode(BaseModel):
    id: str
    type: NodeType
    label: str
    metadata: Dict[str, Any] = Field(default_factory=dict)


class InvestigationEdge(BaseModel):
    id: str
    source: str
    target: str
    relationship: str
    strength: float = Field(..., description="0.0 to 1.0 confidence or strength of correlation")
    evidence: List[str] = Field(default_factory=list)


class InvestigationGraphSummary(BaseModel):
    primary_service: Optional[str] = None
    primary_resource: Optional[str] = None
    primary_region: Optional[str] = None
    primary_event: Optional[str] = None
    strongest_signal: str
    confidence: float
    evidence_count: int
    node_count: int
    edge_count: int


class InvestigationGraphResponse(BaseModel):
    anomaly: AnomalyItemSchema
    nodes: List[InvestigationNode]
    edges: List[InvestigationEdge]
    summary: InvestigationGraphSummary
